code/weapon/trails.h

```diff
@@ -63,7 +66,7 @@ void trail_render_all();
 
 // Returns -1 if failed
 trail *trail_create(trail_info *info);
-void trail_add_segment( trail *trailp, vec3d *pos );
+void trail_add_segment( trail *trailp, vec3d *pos , const matrix* orient = nullptr );
```

Comment:
The only thing I see is that this function now has a default parameter, so I'm assuming there were other places this is called that are not part of this pr?

Reply:
Every instance of `trail_add_segment()` can, was updated to use, and tested with trail spread, so currently all of them send in orient, so it _could_ be made mandatory, but I thought this was a nicer approach in general/for the future since it isn't actually necessary, it would just be a normal trail.

Reply:
I think it's better to have it mandatory now, and then coders in the future could make it optional, when they need it to be. Optional values are nice, but your indirectly communicating to future coders that there already *is* an instance of this function that uses the default value somewhere and they're just not good at looking for it.

The videos do look really nice, btw.

Reply:
@JohnAFernandez I read your comment several times and I'm still not sure what you mean.  If the parameter *can* be optional, and is *designed* to be optional, then why not make it optional?  Punting the "optional" implementation to the future is just going to make it get lost.

The optional form of the function is perfectly reasonable.  Who uses a function without looking at the function prototype?

Reply:
But the optional argument is never used.

Reply:
I'm ambivalent either way. It is not required in order to function like a basic trail. Every caller is capable of using it as a 'spreading' as opposed to basic trail, and so all callers send in the orient (although it may not be used). I'm happy to go with whatever you guys think is a better approach.

Reply:
Let me put it this way:

A rookie coder is looking at this function and is trying to figure out where it's used to get a better handle on what it's used for.  Maybe they have MSVC and intellisense but they're not sure how well they can trust it, since they are brand new.  They search for the function using multiple methods and yet cannot find anywhere using the default argument. 

"Why would they have a default argument, if it is never used?  I must be doing something wrong."

It's just clearer code to not have the default argument, and it makes it more accessible to newer coders, who we shouldn't be discouraging.

Reply:
I think a coder is just as likely to blame the *previous* coder as he is to blame himself.  Even so, this isn't an API problem but a documentation problem.

If there's a good reason to make the argument optional, then make it optional.  If there's a good reason to always require the argument, then always require the argument.

But I think the root problem here can be solved with some documentation.  @Baezon, add some comments explaining what the new argument does, what values are allowed, and why a coder may want to choose one over the other.

Reply:
To be clear, I meant that the argument just needs to not be default.  It is a good idea to add more comments.

Reply:
I do think, for any possible future usage of this I can think of, some forward vector to spread away from orthogonally should be available, so mandatory should be OK. With some documentation explaining this is the purpose of the orient, of course.

Reply:
That sounds good to me. And the documentation is a good start, but please expand a bit on what "spread" is (a link or reference is fine), describe what the orient indicates, and describe how the trail changes with an orient vs. a nullptr.

Reply:
I mean, the function leads them directly to `trail->info.spread` if they want the full explanation, and I'm not sure how much more there is to say about how the orient is used.

Reply:
The comment is confusing.  I tried to make a suggestion but it really wasn't clear what you were trying to say.  The terms need to be made explicit and the explanation needs to provide more detail.